## Objective: 

You are a cab rental start-up company. You have successfully run the pilot project and now want to launch your cab service across the country. You have collected the historical data from your pilot project and now have a requirement to apply analytics for fare prediction. You need to design a system that predicts the fare amount for a cab ride in the city.

In [1]:
#Importing required libraries
import os #getting access to input files
import pandas as pd # Importing pandas for performing EDA
import numpy as np  # Importing numpy for Linear Algebric operations
import matplotlib.pyplot as plt # Importing for Data Visualization
import seaborn as sns # Importing for Data Visualization
from collections import Counter 
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.linear_model import LinearRegression #ML algorithm
from sklearn.model_selection import train_test_split #splitting dataset
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from pprint import pprint
from sklearn.model_selection import GridSearchCV    

%matplotlib inline

In [2]:
#Setting the working directory

os.chdir("E:/edWisor/Cab Fare Prediction")
print(os.getcwd())

E:\edWisor\Cab Fare Prediction


In [3]:
#Loading the data:
train  = pd.read_csv("train_cab.csv",na_values={"pickup_datetime":"43"})
test   = pd.read_csv("test.csv")

## Understanding the data :

In [4]:
train.head() #checking first five rows of the training dataset

fare_amount          pickup_datetime  pickup_longitude  pickup_latitude  \
0         4.5  2009-06-15 17:26:21 UTC        -73.844311        40.721319   
1        16.9  2010-01-05 16:52:16 UTC        -74.016048        40.711303   
2         5.7  2011-08-18 00:35:00 UTC        -73.982738        40.761270   
3         7.7  2012-04-21 04:30:42 UTC        -73.987130        40.733143   
4         5.3  2010-03-09 07:51:00 UTC        -73.968095        40.768008   

   dropoff_longitude  dropoff_latitude  passenger_count  
0         -73.841610         40.712278              1.0  
1         -73.979268         40.782004              1.0  
2         -73.991242         40.750562              2.0  
3         -73.991567         40.758092              1.0  
4         -73.956655         40.783762              1.0

In [5]:
test.head() #checking first five rows of the test dataset

Unnamed: 0      pickup_datetime  pickup_longitude  pickup_latitude  \
0           0  2015-01-27 13:08:24        -73.973320        40.763805   
1           1  2015-01-27 13:08:24        -73.986862        40.719383   
2           2  2011-10-08 11:53:44        -73.982524        40.751260   
3           3  2012-12-01 21:12:12        -73.981160        40.767807   
4           4  2012-12-01 21:12:12        -73.966046        40.789775   

   dropoff_longitude  dropoff_latitude  passenger_count  year  Month  Date  \
0         -73.981430         40.743835                1  2015      1    27   
1         -73.998886         40.739201                1  2015      1    27   
2         -73.979654         40.746139                1  2011     10     8   
3         -73.990448         40.751635                1  2012     12     1   
4         -73.988565         40.744427                1  2012     12     1   

   Day  Hour  Minute  distance  
0    1    13       8  2.323259  
1    1    13       8  2.425353  
2    5    11      53  0.618628  
3    5    21      12  1.961033  
4    5    21      12  5.387301

In [6]:
print("shape of training data is: ",train.shape) #checking the number of rows and columns in training data
print("shape of test data is: ",test.shape) #checking the number of rows and columns in test data

shape of training data is:  (16067, 7)
shape of test data is:  (9914, 14)


In [7]:
train.dtypes #checking the data-types in training dataset

fare_amount           object
pickup_datetime       object
pickup_longitude     float64
pickup_latitude      float64
dropoff_longitude    float64
dropoff_latitude     float64
passenger_count      float64
dtype: object

Here we can see pickup datetime and fare amount is of object type. So we need to change the data type of both.

In [8]:
test.dtypes #checking the data-types in test dataset

Unnamed: 0             int64
pickup_datetime       object
pickup_longitude     float64
pickup_latitude      float64
dropoff_longitude    float64
dropoff_latitude     float64
passenger_count        int64
year                   int64
Month                  int64
Date                   int64
Day                    int64
Hour                   int64
Minute                 int64
distance             float64
dtype: object

In [9]:
train.describe() 

pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
count      16067.000000     16067.000000       16067.000000      16067.000000   
mean         -72.462787        39.914725         -72.462328         39.897906   
std           10.578384         6.826587          10.575062          6.187087   
min          -74.438233       -74.006893         -74.429332        -74.006377   
25%          -73.992156        40.734927         -73.991182         40.734651   
50%          -73.981698        40.752603         -73.980172         40.753567   
75%          -73.966838        40.767381         -73.963643         40.768013   
max           40.766125       401.083332          40.802437         41.366138   

       passenger_count  
count     16012.000000  
mean          2.625070  
std          60.844122  
min           0.000000  
25%           1.000000  
50%           1.000000  
75%           2.000000  
max        5345.000000

In [10]:
test.describe()

Unnamed: 0  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  9914.000000       9914.000000      9914.000000        9914.000000   
mean   4956.500000        -73.974722        40.751041         -73.973657   
std    2862.069618          0.042774         0.033541           0.039072   
min       0.000000        -74.252193        40.573143         -74.263242   
25%    2478.250000        -73.992501        40.736125         -73.991247   
50%    4956.500000        -73.982326        40.753051         -73.980015   
75%    7434.750000        -73.968013        40.767113         -73.964059   
max    9913.000000        -72.986532        41.709555         -72.990963   

       dropoff_latitude  passenger_count         year        Month  \
count       9914.000000      9914.000000  9914.000000  9914.000000   
mean          40.751743         1.671273  2011.815816     6.857979   
std            0.035435         1.278747     1.803347     3.353272   
min           40.568973         1.000000  2009.000000     1.000000   
25%           40.735254         1.000000  2010.000000     4.000000   
50%           40.754065         1.000000  2012.000000     7.000000   
75%           40.768757         2.000000  2014.000000    10.000000   
max           41.696683         6.000000  2015.000000    12.000000   

              Date          Day         Hour       Minute     distance  
count  9914.000000  9914.000000  9914.000000  9914.000000  9914.000000  
mean     16.194170     2.852834    13.467420    29.548416     3.435371  
std       8.838482     1.994451     6.868584    18.674818     3.972374  
min       1.000000     0.000000     0.000000     0.000000     0.000000  
25%       9.000000     1.000000     8.000000    13.000000     1.298277  
50%      16.000000     3.000000    15.000000    33.000000     2.217412  
75%      25.000000     5.000000    19.000000    45.000000     4.045302  
max      31.000000     6.000000    23.000000    59.000000    99.996040

## Data Cleaning & Missing Value Analysis :

In [11]:
#Convert fare_amount from object to numeric
train["fare_amount"] = pd.to_numeric(train["fare_amount"],errors = "coerce")  #Using errors=’coerce’. It will replace all non-numeric values with NaN.

In [12]:
train.dtypes

fare_amount          float64
pickup_datetime       object
pickup_longitude     float64
pickup_latitude      float64
dropoff_longitude    float64
dropoff_latitude     float64
passenger_count      float64
dtype: object

In [13]:
train.shape

(16067, 7)

In [14]:
train.dropna(subset= ["pickup_datetime"])   #dropping NA values in datetime column

fare_amount          pickup_datetime  pickup_longitude  \
0              4.5  2009-06-15 17:26:21 UTC        -73.844311   
1             16.9  2010-01-05 16:52:16 UTC        -74.016048   
2              5.7  2011-08-18 00:35:00 UTC        -73.982738   
3              7.7  2012-04-21 04:30:42 UTC        -73.987130   
4              5.3  2010-03-09 07:51:00 UTC        -73.968095   
...            ...                      ...               ...   
16062          6.5  2014-12-12 07:41:00 UTC        -74.008820   
16063         16.1  2009-07-13 07:58:00 UTC        -73.981310   
16064          8.5  2009-11-11 11:19:07 UTC        -73.972507   
16065          8.1  2010-05-11 23:53:00 UTC        -73.957027   
16066          8.5  2011-12-14 06:24:33 UTC        -74.002111   

       pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  
0            40.721319         -73.841610         40.712278              1.0  
1            40.711303         -73.979268         40.782004              1.0  
2            40.761270         -73.991242         40.750562              2.0  
3            40.733143         -73.991567         40.758092              1.0  
4            40.768008         -73.956655         40.783762              1.0  
...                ...                ...               ...              ...  
16062        40.718757         -73.998865         40.719987              1.0  
16063        40.781695         -74.014392         40.715527              2.0  
16064        40.753417         -73.979577         40.765495              1.0  
16065        40.765945         -73.981983         40.779560              1.0  
16066        40.729755         -73.983877         40.761975              NaN  

[16066 rows x 7 columns]

In [15]:
# Here pickup_datetime variable is in object so we need to change its data type to datetime
train['pickup_datetime'] =  pd.to_datetime(train['pickup_datetime'], format='%Y-%m-%d %H:%M:%S UTC')

In [16]:
### we will saperate the Pickup_datetime column into separate field like year, month, day of the week, etc

train['year'] = train['pickup_datetime'].dt.year
train['Month'] = train['pickup_datetime'].dt.month
train['Date'] = train['pickup_datetime'].dt.day
train['Day'] = train['pickup_datetime'].dt.dayofweek
train['Hour'] = train['pickup_datetime'].dt.hour
train['Minute'] = train['pickup_datetime'].dt.minute

In [17]:
train.dtypes #Re-checking datatypes after conversion

fare_amount                 float64
pickup_datetime      datetime64[ns]
pickup_longitude            float64
pickup_latitude             float64
dropoff_longitude           float64
dropoff_latitude            float64
passenger_count             float64
year                        float64
Month                       float64
Date                        float64
Day                         float64
Hour                        float64
Minute                      float64
dtype: object

In [18]:
test["pickup_datetime"] = pd.to_datetime(test["pickup_datetime"],format= "%Y-%m-%d %H:%M:%S UTC")

ValueError: time data '2015-01-27 13:08:24' does not match format '%Y-%m-%d %H:%M:%S UTC' (match)

In [ ]:
### we will saperate the Pickup_datetime column into separate field like year, month, day of the week, etc

test['year'] = test['pickup_datetime'].dt.year
test['Month'] = test['pickup_datetime'].dt.month
test['Date'] = test['pickup_datetime'].dt.day
test['Day'] = test['pickup_datetime'].dt.dayofweek
test['Hour'] = test['pickup_datetime'].dt.hour
test['Minute'] = test['pickup_datetime'].dt.minute

In [ ]:
test.dtypes #Re-checking test datatypes after conversion

#### Observations :

1. An outlier in pickup_datetime column of value 43
2. Passenger count should not exceed 6(even if we consider SUV)
3. Latitudes range from -90 to 90. Longitudes range from -180 to 180
4. Few missing values and High values of fare and Passenger count are present. So, decided to remove them.

Checking the Datetime Variable : 

In [ ]:
#removing datetime missing values rows
train = train.drop(train[train['pickup_datetime'].isnull()].index, axis=0)
print(train.shape)
print(train['pickup_datetime'].isnull().sum())

Checking the passenger count variable : 

In [ ]:
train["passenger_count"].describe()

We can see maximum number of passanger count is 5345 which is actually not possible. So reducing the passenger count to 6 (even if we consider the SUV)

In [ ]:
train = train.drop(train[train["passenger_count"]> 6 ].index, axis=0)

In [ ]:
#Also removing the values with passenger count of 0.
train = train.drop(train[train["passenger_count"] == 0 ].index, axis=0)

In [ ]:
train["passenger_count"].describe()

In [ ]:
train["passenger_count"].sort_values(ascending= True)

In [ ]:
#removing passanger_count missing values rows
train = train.drop(train[train['passenger_count'].isnull()].index, axis=0)
print(train.shape)
print(train['passenger_count'].isnull().sum())

There is one passenger count value of 0.12 which is not possible. Hence we will remove fractional passenger value

In [ ]:
train = train.drop(train[train["passenger_count"] == 0.12 ].index, axis=0)
train.shape

Next checking the Fare Amount variable :

In [ ]:
##finding decending order of fare to get to know whether the outliers are present or not
train["fare_amount"].sort_values(ascending=False)

In [ ]:
Counter(train["fare_amount"]<0)

In [ ]:
train = train.drop(train[train["fare_amount"]<0].index, axis=0)
train.shape

In [ ]:
##make sure there is no negative values in the fare_amount variable column
train["fare_amount"].min()

In [ ]:
#Also remove the row where fare amount is zero
train = train.drop(train[train["fare_amount"]<1].index, axis=0)
train.shape

In [ ]:
#Now we can see that there is a huge difference in 1st 2nd and 3rd position in decending order of fare amount
# so we will remove the rows having fare amounting more that 454 as considering them as outliers

train = train.drop(train[train["fare_amount"]> 454 ].index, axis=0)
train.shape

In [ ]:
# eliminating rows for which value of "fare_amount" is missing
train = train.drop(train[train['fare_amount'].isnull()].index, axis=0)
print(train.shape)
print(train['fare_amount'].isnull().sum())

In [ ]:
train["fare_amount"].describe()

Now checking the pickup lattitude and longitude :

In [ ]:
#Lattitude----(-90 to 90)
#Longitude----(-180 to 180)

# we need to drop the rows having  pickup lattitute and longitute out the range mentioned above

#train = train.drop(train[train['pickup_latitude']<-90])
train[train['pickup_latitude']<-90]
train[train['pickup_latitude']>90]

In [ ]:
#Hence dropping one value of >90
train = train.drop((train[train['pickup_latitude']<-90]).index, axis=0)
train = train.drop((train[train['pickup_latitude']>90]).index, axis=0)


In [ ]:
train[train['pickup_longitude']<-180]
train[train['pickup_longitude']>180]

In [ ]:
train[train['dropoff_latitude']<-90]
train[train['dropoff_latitude']>90]

In [ ]:
train[train['dropoff_longitude']<-180]
train[train['dropoff_longitude']>180]

In [ ]:
train.shape

In [ ]:
train.isnull().sum()

In [ ]:
test.isnull().sum()

### Now we have successfully cleared our both datasets. Thus proceeding for further operations:

Calculating distance based on the given coordinates :


In [ ]:
#As we know that we have given pickup longitute and latitude values and same for drop. 
#So we need to calculate the distance Using the haversine formula and we will create a new variable called distance
from math import radians, cos, sin, asin, sqrt

def haversine(a):
    lon1=a[0]
    lat1=a[1]
    lon2=a[2]
    lat2=a[3]
    """
    Calculate the great circle distance between two points 
    on the earth (specified in decimal degrees)
    """
    # convert decimal degrees to radians 
    lon1, lat1, lon2, lat2 = map(radians, [lon1, lat1, lon2, lat2])
    
    # haversine formula 
    dlon = lon2 - lon1 
    dlat = lat2 - lat1 
    a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
    c =  2 * asin(sqrt(a))
    # Radius of earth in kilometers is 6371
    km = 6371* c
    return km
# 1min 

In [ ]:
train['distance'] = train[['pickup_longitude','pickup_latitude','dropoff_longitude','dropoff_latitude']].apply(haversine,axis=1)

In [ ]:
test['distance'] = test[['pickup_longitude','pickup_latitude','dropoff_longitude','dropoff_latitude']].apply(haversine,axis=1)

In [ ]:
train.head()

In [ ]:
test.head()

In [ ]:
train.nunique()

In [ ]:
test.nunique()

In [ ]:
##finding decending order of fare to get to know whether the outliers are presented or not
train['distance'].sort_values(ascending=False)

As we can see that top 23 values in the distance variables are very high It means more than 8000 Kms distance they have travelled
Also just after 23rd value from the top, the distance goes down to 127, which means these values are showing some outliers
We need to remove these values


In [ ]:
Counter(train['distance'] == 0)

In [ ]:
Counter(test['distance'] == 0)

In [ ]:
Counter(train['fare_amount'] == 0)

In [ ]:
###we will remove the rows whose distance value is zero

train = train.drop(train[train['distance']== 0].index, axis=0)
train.shape

In [ ]:
#we will remove the rows whose distance values is very high which is more than 129kms
train = train.drop(train[train['distance'] > 130 ].index, axis=0)
train.shape

In [ ]:
train.head()

Now we have splitted the pickup date time variable into different varaibles like month, year, day etc so now we dont need to have that pickup_Date variable now. Hence we can drop that, Also we have created distance using pickup and drop longitudes and latitudes so we will also drop pickup and drop longitudes and latitudes variables.

In [ ]:
drop = ['pickup_datetime', 'pickup_longitude', 'pickup_latitude','dropoff_longitude', 'dropoff_latitude', 'Minute']
train = train.drop(drop, axis = 1)

In [ ]:
train.head()

In [ ]:
train['passenger_count'] = train['passenger_count'].astype('int64')
train['year'] = train['year'].astype('int64')
train['Month'] = train['Month'].astype('int64')
train['Date'] = train['Date'].astype('int64')
train['Day'] = train['Day'].astype('int64')
train['Hour'] = train['Hour'].astype('int64')

In [ ]:
train.dtypes

In [ ]:
drop_test = ['pickup_datetime', 'pickup_longitude', 'pickup_latitude','dropoff_longitude', 'dropoff_latitude', 'Minute']
test = test.drop(drop_test, axis = 1)

In [ ]:
test.head()

In [ ]:
test.dtypes

# Data Visualization :

Visualization of following:

1. Number of Passengers effects the the fare
2. Pickup date and time effects the fare
3. Day of the week does effects the fare
4. Distance effects the fare

In [ ]:
# Count plot on passenger count
plt.figure(figsize=(15,7))
sns.countplot(x="passenger_count", data=train)

In [ ]:
#Relationship beetween number of passengers and Fare

plt.figure(figsize=(15,7))
plt.scatter(x=train['passenger_count'], y=train['fare_amount'], s=10)
plt.xlabel('No. of Passengers')
plt.ylabel('Fare')
plt.show()

### Observations :
   By seeing the above plots we can easily conclude that:
1. single travelling passengers are most frequent travellers.
2. At the sametime we can also conclude that highest Fare are coming from single & double travelling passengers.

In [ ]:
#Relationship between date and Fare
plt.figure(figsize=(15,7))
plt.scatter(x=train['Date'], y=train['fare_amount'], s=10)
plt.xlabel('Date')
plt.ylabel('Fare')
plt.show()

In [ ]:
plt.figure(figsize=(15,7))
train.groupby(train["Hour"])['Hour'].count().plot(kind="bar")
plt.show()

Lowest cabs at 5 AM and highest at and around 7 PM i.e the office rush hours

In [ ]:
#Relationship between Time and Fare
plt.figure(figsize=(15,7))
plt.scatter(x=train['Hour'], y=train['fare_amount'], s=10)
plt.xlabel('Hour')
plt.ylabel('Fare')
plt.show()

From the above plot We can observe that the cabs taken at 7 am and 23 Pm are the costliest. 
Hence we can assume that cabs taken early in morning and late at night are costliest


In [ ]:
#impact of Day on the number of cab rides
plt.figure(figsize=(15,7))
sns.countplot(x="Day", data=train)

Observation :
The day of the week does not seem to have much influence on the number of cabs ride

In [ ]:
#Relationships between day and Fare
plt.figure(figsize=(15,7))
plt.scatter(x=train['Day'], y=train['fare_amount'], s=10)
plt.xlabel('Day')
plt.ylabel('Fare')
plt.show()

The highest fares seem to be on a Sunday, Monday and Thursday, and the low on Wednesday and Saturday. May be due to low demand of the cabs on saturdays the cab fare is low and high demand of cabs on sunday and monday shows the high fare prices

In [ ]:
#Relationship between distance and fare 
plt.figure(figsize=(15,7))
plt.scatter(x = train['distance'],y = train['fare_amount'],c = "g")
plt.xlabel('Distance')
plt.ylabel('Fare')
plt.show()

It is quite obvious that distance will effect the amount of fare

# Feature Scaling :

In [ ]:
#Normality check of training data is uniformly distributed or not-

for i in ['fare_amount', 'distance']:
    print(i)
    sns.distplot(train[i],bins='auto',color='green')
    plt.title("Distribution for Variable "+i)
    plt.ylabel("Density")
    plt.show()

In [ ]:
#since skewness of target variable is high, apply log transform to reduce the skewness-
train['fare_amount'] = np.log1p(train['fare_amount'])

#since skewness of distance variable is high, apply log transform to reduce the skewness-
train['distance'] = np.log1p(train['distance'])

In [ ]:
#Normality Re-check to check data is uniformly distributed or not after log transformartion

for i in ['fare_amount', 'distance']:
    print(i)
    sns.distplot(train[i],bins='auto',color='green')
    plt.title("Distribution for Variable "+i)
    plt.ylabel("Density")
    plt.show()

Here we can see bell shaped distribution. Hence our continous variables are now normally distributed, we will use not use any  Feature Scalling technique. i.e, Normalization or Standarization for our training data

In [ ]:
#Normality check for test data is uniformly distributed or not-

sns.distplot(test['distance'],bins='auto',color='green')
plt.title("Distribution for Variable "+i)
plt.ylabel("Density")
plt.show()

In [ ]:
#since skewness of distance variable is high, apply log transform to reduce the skewness-
test['distance'] = np.log1p(test['distance'])

In [ ]:
#rechecking the distribution for distance
sns.distplot(test['distance'],bins='auto',color='green')
plt.title("Distribution for Variable "+i)
plt.ylabel("Density")
plt.show()


As we can see a bell shaped distribution. Hence our continous variables are now normally distributed, we will use not use any  Feature Scalling technique. i.e, Normalization or Standarization for our test data

# Applying ML ALgorithms: 

In [ ]:
##train test split for further modelling
X_train, X_test, y_train, y_test = train_test_split( train.iloc[:, train.columns != 'fare_amount'], 
                         train.iloc[:, 0], test_size = 0.20, random_state = 1)

In [ ]:
print(X_train.shape)
print(X_test.shape)

### Linear Regression Model :

In [ ]:
# Building model on top of training dataset
fit_LR = LinearRegression().fit(X_train , y_train)

In [ ]:
#prediction on train data
pred_train_LR = fit_LR.predict(X_train)

In [ ]:
#prediction on test data
pred_test_LR = fit_LR.predict(X_test)

In [ ]:
##calculating RMSE for test data
RMSE_test_LR = np.sqrt(mean_squared_error(y_test, pred_test_LR))

##calculating RMSE for train data
RMSE_train_LR= np.sqrt(mean_squared_error(y_train, pred_train_LR))

In [ ]:
print("Root Mean Squared Error For Training data = "+str(RMSE_train_LR))
print("Root Mean Squared Error For Test data = "+str(RMSE_test_LR))

In [ ]:
#calculate R^2 for train data
from sklearn.metrics import r2_score
r2_score(y_train, pred_train_LR)

In [ ]:
r2_score(y_test, pred_test_LR)

### Decision tree Model : 

In [ ]:
fit_DT = DecisionTreeRegressor(max_depth = 2).fit(X_train,y_train)

In [ ]:
#prediction on train data
pred_train_DT = fit_DT.predict(X_train)

#prediction on test data
pred_test_DT = fit_DT.predict(X_test)

In [ ]:
##calculating RMSE for train data
RMSE_train_DT = np.sqrt(mean_squared_error(y_train, pred_train_DT))

##calculating RMSE for test data
RMSE_test_DT = np.sqrt(mean_squared_error(y_test, pred_test_DT))

In [ ]:
print("Root Mean Squared Error For Training data = "+str(RMSE_train_DT))
print("Root Mean Squared Error For Test data = "+str(RMSE_test_DT))

In [ ]:
## R^2 calculation for train data
r2_score(y_train, pred_train_DT)

In [ ]:
## R^2 calculation for test data
r2_score(y_test, pred_test_DT)

### Random Forest Model :

In [ ]:
fit_RF = RandomForestRegressor(n_estimators = 200).fit(X_train,y_train)

In [ ]:
#prediction on train data
pred_train_RF = fit_RF.predict(X_train)
#prediction on test data
pred_test_RF = fit_RF.predict(X_test)

In [ ]:
##calculating RMSE for train data
RMSE_train_RF = np.sqrt(mean_squared_error(y_train, pred_train_RF))
##calculating RMSE for test data
RMSE_test_RF = np.sqrt(mean_squared_error(y_test, pred_test_RF))

In [ ]:
print("Root Mean Squared Error For Training data = "+str(RMSE_train_RF))
print("Root Mean Squared Error For Test data = "+str(RMSE_test_RF))

In [ ]:
## calculate R^2 for train data

r2_score(y_train, pred_train_RF)

In [ ]:
#calculate R^2 for test data
r2_score(y_test, pred_test_RF)

### Gradient Boosting :

In [ ]:
fit_GB = GradientBoostingRegressor().fit(X_train, y_train)

In [ ]:
#prediction on train data
pred_train_GB = fit_GB.predict(X_train)

#prediction on test data
pred_test_GB = fit_GB.predict(X_test)

In [ ]:
##calculating RMSE for train data
RMSE_train_GB = np.sqrt(mean_squared_error(y_train, pred_train_GB))
##calculating RMSE for test data
RMSE_test_GB = np.sqrt(mean_squared_error(y_test, pred_test_GB))

In [ ]:
print("Root Mean Squared Error For Training data = "+str(RMSE_train_GB))
print("Root Mean Squared Error For Test data = "+str(RMSE_test_GB))

In [ ]:
#calculate R^2 for test data
r2_score(y_test, pred_test_GB)

In [ ]:
#calculate R^2 for train data
r2_score(y_train, pred_train_GB)

# Optimizing the results with parameters tuning :

In [ ]:
from sklearn.ensemble import RandomForestRegressor
rf = RandomForestRegressor(random_state = 42)
from pprint import pprint
# Look at parameters used by our current forest
print('Parameters currently in use:\n')
pprint(rf.get_params())

In [ ]:
##Random Hyperparameter Grid


In [ ]:
from sklearn.model_selection import train_test_split,RandomizedSearchCV


In [ ]:
##Random Search CV on Random Forest Model

RRF = RandomForestRegressor(random_state = 0)
n_estimator = list(range(1,20,2))
depth = list(range(1,100,2))

# Create the random grid
rand_grid = {'n_estimators': n_estimator,
               'max_depth': depth}

randomcv_rf = RandomizedSearchCV(RRF, param_distributions = rand_grid, n_iter = 5, cv = 5, random_state=0)
randomcv_rf = randomcv_rf.fit(X_train,y_train)
predictions_RRF = randomcv_rf.predict(X_test)

view_best_params_RRF = randomcv_rf.best_params_

best_model = randomcv_rf.best_estimator_

predictions_RRF = best_model.predict(X_test)

#R^2
RRF_r2 = r2_score(y_test, predictions_RRF)
#Calculating RMSE
RRF_rmse = np.sqrt(mean_squared_error(y_test,predictions_RRF))

print('Random Search CV Random Forest Regressor Model Performance:')
print('Best Parameters = ',view_best_params_RRF)
print('R-squared = {:0.2}.'.format(RRF_r2))
print('RMSE = ',RRF_rmse)

In [ ]:
gb = GradientBoostingRegressor(random_state = 42)
from pprint import pprint
# Look at parameters used by our current forest
print('Parameters currently in use:\n')
pprint(gb.get_params())

In [ ]:
##Random Search CV on gradient boosting model

gb = GradientBoostingRegressor(random_state = 0)
n_estimator = list(range(1,20,2))
depth = list(range(1,100,2))

# Create the random grid
rand_grid = {'n_estimators': n_estimator,
               'max_depth': depth}

randomcv_gb = RandomizedSearchCV(gb, param_distributions = rand_grid, n_iter = 5, cv = 5, random_state=0)
randomcv_gb = randomcv_gb.fit(X_train,y_train)
predictions_gb = randomcv_gb.predict(X_test)

view_best_params_gb = randomcv_gb.best_params_

best_model = randomcv_gb.best_estimator_

predictions_gb = best_model.predict(X_test)

#R^2
gb_r2 = r2_score(y_test, predictions_gb)
#Calculating RMSE
gb_rmse = np.sqrt(mean_squared_error(y_test,predictions_gb))

print('Random Search CV Gradient Boosting Model Performance:')
print('Best Parameters = ',view_best_params_gb)
print('R-squared = {:0.2}.'.format(gb_r2))
print('RMSE = ', gb_rmse)

In [ ]:
from sklearn.model_selection import GridSearchCV    
## Grid Search CV for random Forest model
regr = RandomForestRegressor(random_state = 0)
n_estimator = list(range(11,20,1))
depth = list(range(5,15,2))

# Create the grid
grid_search = {'n_estimators': n_estimator,
               'max_depth': depth}

## Grid Search Cross-Validation with 5 fold CV
gridcv_rf = GridSearchCV(regr, param_grid = grid_search, cv = 5)
gridcv_rf = gridcv_rf.fit(X_train,y_train)
view_best_params_GRF = gridcv_rf.best_params_

#Apply model on test data
predictions_GRF = gridcv_rf.predict(X_test)

#R^2
GRF_r2 = r2_score(y_test, predictions_GRF)
#Calculating RMSE
GRF_rmse = np.sqrt(mean_squared_error(y_test,predictions_GRF))

print('Grid Search CV Random Forest Regressor Model Performance:')
print('Best Parameters = ',view_best_params_GRF)
print('R-squared = {:0.2}.'.format(GRF_r2))
print('RMSE = ',(GRF_rmse))

In [ ]:
## Grid Search CV for gradinet boosting
gb = GradientBoostingRegressor(random_state = 0)
n_estimator = list(range(11,20,1))
depth = list(range(5,15,2))

# Create the grid
grid_search = {'n_estimators': n_estimator,
               'max_depth': depth}

## Grid Search Cross-Validation with 5 fold CV
gridcv_gb = GridSearchCV(gb, param_grid = grid_search, cv = 5)
gridcv_gb = gridcv_gb.fit(X_train,y_train)
view_best_params_Ggb = gridcv_gb.best_params_

#Apply model on test data
predictions_Ggb = gridcv_gb.predict(X_test)

#R^2
Ggb_r2 = r2_score(y_test, predictions_Ggb)
#Calculating RMSE
Ggb_rmse = np.sqrt(mean_squared_error(y_test,predictions_Ggb))

print('Grid Search CV Gradient Boosting regression Model Performance:')
print('Best Parameters = ',view_best_params_Ggb)
print('R-squared = {:0.2}.'.format(Ggb_r2))
print('RMSE = ',(Ggb_rmse))

# Prediction of fare from provided test dataset :

We have already cleaned and processed our test dataset along with our training dataset. Hence we will be predicting using grid search CV for random forest model

In [ ]:
  
## Grid Search CV for random Forest model
regr = RandomForestRegressor(random_state = 0)
n_estimator = list(range(11,20,1))
depth = list(range(5,15,2))

# Create the grid
grid_search = {'n_estimators': n_estimator,
               'max_depth': depth}

## Grid Search Cross-Validation with 5 fold CV
gridcv_rf = GridSearchCV(regr, param_grid = grid_search, cv = 5)
gridcv_rf = gridcv_rf.fit(X_train,y_train)
view_best_params_GRF = gridcv_rf.best_params_

#Apply model on test data
predictions_GRF_test_Df = gridcv_rf.predict(test)

In [ ]:
predictions_GRF_test_Df

In [ ]:
test['Predicted_fare'] = predictions_GRF_test_Df

In [ ]:
test.head()

In [ ]:
test.to_csv('test.csv')